In [1]:
import pandas as pd

listings = pd.read_csv("../data/raw/listings.csv")
calendar = pd.read_csv("../data/raw/calendar.csv")
reviews = pd.read_csv("../data/raw/reviews.csv")
neighbourhoods = pd.read_csv("../data/raw/neighbourhoods.csv")

print("Listings shape:", listings.shape)
print("Calendar shape:", calendar.shape)
print("Reviews shape:", reviews.shape)
print("Neighbourhoods shape:", neighbourhoods.shape)

Listings shape: (30259, 90)
Calendar shape: (11152576, 5)
Reviews shape: (990170, 6)
Neighbourhoods shape: (230, 2)


In [2]:
listings.head()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,...,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,2539,https://www.airbnb.com/rooms/2539,20260614073253,2026-06-15,city scrape,11 Min to Manhattan • Prospect Park • Fast WiFi,"Bright, serene room in a renovated apartment h...",NaN,https://a0.muscache.com/pictures/hosting/Hosti...,2787,...,5.00,4.78,4.78,NaN,NaN,5,0,5,0,0.08
1,6848,https://www.airbnb.com/rooms/6848,20260614073253,2026-06-14,city scrape,Only 2 stops to Manhattan studio,Comfortable studio apartment with super comfor...,NaN,https://a0.muscache.com/pictures/e4f031a7-f146...,15991,...,4.80,4.69,4.59,NaN,NaN,1,1,0,0,0.95
2,6872,https://www.airbnb.com/rooms/6872,20260614073253,2026-06-14,city scrape,Uptown Sanctuary w/ Private Bath (Month to Month),This charming distancing-friendly month-to-mon...,NaN,https://a0.muscache.com/pictures/miso/Hosting-...,16104,...,5.00,5.00,5.00,NaN,NaN,2,0,2,0,0.04
3,6990,https://www.airbnb.com/rooms/6990,20260614073253,2026-06-14,city scrape,UES Beautiful Blue Room,Beautiful peaceful healthy home,NaN,https://a0.muscache.com/pictures/45fb4ec7-6856...,16800,...,4.94,4.85,4.84,NaN,NaN,1,0,1,0,1.24
4,7097,https://www.airbnb.com/rooms/7097,20260614073253,2026-06-14,city scrape,"Perfect for Your Parents, With Garden & Patio",Parents/grandparents coming to town or are you...,NaN,https://a0.muscache.com/pictures/aaac19fc-4b4d...,17571,...,4.93,4.95,4.82,NaN,NaN,2,0,2,0,2.12


In [3]:
listings.info()

<class 'pandas.DataFrame'>
RangeIndex: 30259 entries, 0 to 30258
Data columns (total 90 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            30259 non-null  int64  
 1   listing_url                                   30259 non-null  str    
 2   scrape_id                                     30259 non-null  int64  
 3   last_scraped                                  30259 non-null  str    
 4   source                                        30259 non-null  str    
 5   name                                          30258 non-null  str    
 6   description                                   29357 non-null  str    
 7   neighborhood_overview                         0 non-null      float64
 8   picture_url                                   30259 non-null  str    
 9   host_id                                       30259 non-null  int64  
 1

In [4]:
# Section 3.1: Data Ingestion & Profiling
# Task: Identify missing values and data quality issues

missing_values = listings.isnull().sum()

missing_report = (
    missing_values[missing_values > 0]
    .sort_values(ascending=False)
)

missing_report.head(20)

neighborhood_overview          30259
host_since                     30259
host_neighbourhood             30259
host_verifications             30259
host_total_listings_count      30259
host_thumbnail_url             30259
host_response_time             30259
host_acceptance_rate           30259
host_response_rate             30259
calendar_updated               30259
instant_bookable               30259
neighbourhood                  30259
license                        24973
host_about                     12410
bedrooms                       10973
bathrooms                      10112
beds                            9420
price_quote_price_per_night     8745
price_quote_total_price         8745
estimated_revenue_l365d         8744
dtype: int64

In [5]:
# Section 3.1: Data Ingestion & Profiling
# Task: Generate descriptive statistics

listings.describe()

,id,scrape_id,neighborhood_overview,host_id,host_profile_id,host_since,hosts_time_as_user_years,hosts_time_as_user_months,hosts_time_as_host_years,hosts_time_as_host_months,...,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
count,3.025900e+04,3.025900e+04,0.0,3.025900e+04,2.991000e+04,0.0,29910.000000,29910.000000,29910.000000,29910.000000,...,21692.000000,21697.000000,21691.000000,21691.000000,0.0,30259.000000,30259.000000,30259.000000,30259.000000,21700.000000
mean,6.172513e+17,2.026061e+13,NaN,2.672815e+15,1.469574e+18,NaN,8.284754,5.653928,6.684286,5.631829,...,4.838066,4.821284,4.743640,4.633094,NaN,55.887141,45.630721,9.146601,0.095905,0.846505
std,6.155147e+17,0.000000e+00,NaN,6.705743e+16,2.417060e+16,NaN,3.934796,3.482520,3.889389,3.356700,...,0.373301,0.418090,0.388374,0.496439,NaN,174.246090,172.331916,36.412035,1.125182,1.847935
min,2.539000e+03,2.026061e+13,NaN,1.678000e+03,1.462506e+18,NaN,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,NaN,1.000000,0.000000,0.000000,0.000000,0.010000
25%,3.038760e+07,2.026061e+13,NaN,2.069159e+07,1.462882e+18,NaN,5.000000,3.000000,3.000000,3.000000,...,4.827500,4.820000,4.660000,4.530000,NaN,1.000000,0.000000,0.000000,0.000000,0.090000
50%,6.608782e+17,2.026061e+13,NaN,1.074344e+08,1.465524e+18,NaN,9.000000,6.000000,7.000000,5.000000,...,4.950000,4.960000,4.850000,4.760000,NaN,2.000000,1.000000,0.000000,0.000000,0.270000
75%,1.160901e+18,2.026061e+13,NaN,3.948700e+08,1.469572e+18,NaN,11.000000,9.000000,10.000000,8.000000,...,5.000000,5.000000,5.000000,4.920000,NaN,12.000000,3.000000,2.000000,0.000000,0.962500
max,1.707075e+18,2.026061e+13,NaN,1.704453e+18,1.706707e+18,NaN,17.000000,11.000000,15.000000,11.000000,...,5.000000,5.000000,5.000000,5.000000,NaN,955.000000,955.000000,231.000000,22.000000,115.630000


In [6]:
# Section 2.3: Exploration Tasks
# Task: Document dataset schema and columns

print(listings.columns.tolist())

['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name', 'description', 'neighborhood_overview', 'picture_url', 'host_id', 'host_url', 'host_profile_id', 'host_profile_url', 'host_name', 'host_since', 'hosts_time_as_user_years', 'hosts_time_as_user_months', 'hosts_time_as_host_years', 'hosts_time_as_host_months', 'host_location', 'host_about', 'host_response_time', 'host_response_rate', 'host_acceptance_rate', 'host_is_superhost', 'host_thumbnail_url', 'host_picture_url', 'host_neighbourhood', 'host_listings_count', 'host_total_listings_count', 'host_verifications', 'host_has_profile_pic', 'host_identity_verified', 'neighbourhood', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude', 'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms', 'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price', 'price_quote_checkin_date', 'price_quote_checkout_date', 'price_quote_total_price', 'price_quote_price_per_night', 'price_quote_raw', 'minimum_nig

In [7]:
# Section 3.1: Data Ingestion & Profiling
# Task: Profile dataset dimensions and data types

print("\nListings Data Types:\n")
print(listings.dtypes.value_counts())

print("\nListings Memory Usage:")
print(
    round(
        listings.memory_usage(deep=True).sum() / 1024**2,
        2
    ),
    "MB"
)


Listings Data Types:

float64    42
str        30
int64      18
Name: count, dtype: int64

Listings Memory Usage:
128.08 MB


## Data Quality Observations

- Several columns are completely empty and cannot be used for analysis.
- Key listing attributes such as bedrooms, bathrooms, and beds contain missing values.
- License information is unavailable for most listings.
- Revenue-related fields require further investigation before use.

In [8]:
# Section 3.1: Data Ingestion & Profiling
# Task: Identify columns with 100% missing values

fully_missing = listings.columns[listings.isnull().all()]

print("Columns with 100% missing values:\n")

for column in fully_missing:
    print(column)

Columns with 100% missing values:

neighborhood_overview
host_since
host_response_time
host_response_rate
host_acceptance_rate
host_thumbnail_url
host_neighbourhood
host_total_listings_count
host_verifications
neighbourhood
calendar_updated
instant_bookable


In [9]:
# Section 3.1: Data Ingestion & Profiling
# Task: Detect duplicate records

duplicates = listings.duplicated().sum()

print(f"Duplicate rows: {duplicates}")

Duplicate rows: 0


In [10]:
# Section 3.1: Data Ingestion & Profiling
# Task: Validate business rules

negative_prices = (listings["price"] < 0).sum()

print(f"Listings with negative prices: {negative_prices}")

TypeError: '<' not supported between instances of 'str' and 'int'

In [11]:
# Section 3.1: Data Ingestion & Profiling
# Task: Inspect the price column

print(listings["price"].head())
print(listings["price"].dtype)

0    $113.97
1    $117.27
2     $80.06
3     $77.17
4    $202.47
Name: price, dtype: str
str


In [12]:
# Section 3.2: Data Cleaning & Standardization
# Task: Standardize the price column

listings["price"] = (
    listings["price"]
    .astype(str)
    .str.replace("$", "", regex=False)
    .str.replace(",", "", regex=False)
    .astype(float)
)

print(listings["price"].head())
print(listings["price"].dtype)

0    113.97
1    117.27
2     80.06
3     77.17
4    202.47
Name: price, dtype: float64
float64


In [13]:
# Section 3.1: Data Ingestion & Profiling
# Task: Validate business rules (rerun after cleaning)

negative_prices = (listings["price"] < 0).sum()

print(f"Listings with negative prices: {negative_prices}")

Listings with negative prices: 0


In [14]:
# Section 3.2: Data Cleaning & Standardization
# Task: Standardize date fields

listings["last_scraped"] = pd.to_datetime(listings["last_scraped"])
calendar["date"] = pd.to_datetime(calendar["date"])
reviews["date"] = pd.to_datetime(reviews["date"])

In [15]:
# Section 3.2: Data Cleaning & Standardization
# Task: Verify date conversions

print(listings["last_scraped"].dtype)
print(calendar["date"].dtype)
print(reviews["date"].dtype)

datetime64[us]
datetime64[us]
datetime64[us]


In [16]:
# Section 2.3: Exploration Tasks
# Task: Inspect key columns

important_columns = [
    "id",
    "host_id",
    "price",
    "room_type",
    "property_type",
    "neighbourhood_cleansed",
    "number_of_reviews",
    "review_scores_rating"
]

listings[important_columns].head()

,id,host_id,price,room_type,property_type,neighbourhood_cleansed,number_of_reviews,review_scores_rating
0,2539,2787,113.97,Private room,Private room in condo,Kensington,10,4.80
1,6848,15991,117.27,Entire home/apt,Entire rental unit,Williamsburg,198,4.60
2,6872,16104,80.06,Private room,Private room in condo,East Harlem,2,5.00
3,6990,16800,77.17,Private room,Private room in rental unit,East Harlem,251,4.88
4,7097,17571,202.47,Private room,Private room in guest suite,Fort Greene,423,4.89


# Section Summary: Dataset Familiarization and Profiling

## Completed Tasks

- Explored the Inside Airbnb datasets for New York City.
- Documented the dimensions of each dataset.
- Identified missing values and data quality issues.
- Examined the schema and data types.
- Validated business rules on key fields.

## Key Findings

- The listings dataset contains 30,259 records and 90 columns.
- The calendar dataset contains more than 11 million records.
- Several columns are completely missing and require investigation.
- The price column required cleaning before analysis.

## Challenges

- Large calendar data may require optimization techniques.
- Missing values are present in several important columns.